mining.go

```diff
@@ -695,7 +709,7 @@ mempoolLoop:
 				// one.
 				item := e.Value.(*txPrioItem)
 				delete(item.dependsOn, *tx.Sha())
-				if len(item.dependsOn) == 0 {
+				if len(item.dependsOn) == 0 && !priorityQueue.Contains(item) {
```

Comment:
I think looping the entire priority queue here will cause performance issues.  Keep in mind that there often by 30,000+ transaction in the pq after the node has been running for a while.  Due to that, I believe it would be a lot more efficient to simply keep track of the hashes of the added transactions (priority queue items) in a map and avoid adding them that way.

The following diff should do the trick and remove the need for the `Contains` function as well:

```diff
diff --git a/mining.go b/mining.go
index ccf3d34..5955931 100644
--- a/mining.go
+++ b/mining.go
@@ -689,14 +689,22 @@ mempoolLoop:
 		// have any other unsatisified dependencies) to the priority
 		// queue.
 		if deps != nil {
+			pushedTxns := make(map[wire.ShaHash]struct{})
 			for e := deps.Front(); e != nil; e = e.Next() {
 				// Add the transaction to the priority queue if
 				// there are no more dependencies after this
-				// one.
+				// one and it wasn't already added via another
+				// dependency.
 				item := e.Value.(*txPrioItem)
 				delete(item.dependsOn, *tx.Sha())
 				if len(item.dependsOn) == 0 {
+					itemHash := item.tx.Sha()
+					if _, ok := pushedTxns[*itemHash]; ok {
+						continue
+					}
+
 					heap.Push(priorityQueue, item)
+					pushedTxns[*itemHash] = struct{}{}
 				}
 			}
 		}
```